### PR TITLE
fix: use provider-declared schemes for version ordering

### LINKED ISSUE #306: Version ordering is inconsistent across provider formats and application flows
## Problem

GenHub compares content versions in multiple update, download, profile, and reconciliation paths. These paths can apply different rules to the same provider version.

Providers also use different formats:

- Community Outpost uses ISO dates.
- Generals Online uses `MMDDYY_QFE#` with optional build tags.
- TheSuperHackers uses numeric, date-stamp, and prefixed versions.

A concrete Generals Online failure occurs across year boundaries:

- Installed: `121525_QFE1` — December 15, 2025
- Available: `060526_QFE1` — June 5, 2026

June 2026 is newer, but integer or raw-string ordering can treat it as older. Build tags can also incorrectly affect release precedence, and separate call sites can disagree about the newest installed version.

## Expected behavior

- Each provider declares the version scheme used for ordering.
- Update, download, profile, and reconciliation flows use one shared provider-aware comparison path.
- Generals Online releases order chronologically by year, month, day, then QFE.
- Build tags do not affect release precedence.
- Malformed versions are handled deterministically and culture-invariantly.
- Persistent manifest identity remains separate from release ordering.

## Acceptance criteria

- Provider definitions select the applicable numeric, ISO-date, or MMDDYY-QFE scheme.
- Relevant application flows use the shared comparer.
- Cross-year Generals Online versions and QFE revisions order correctly.
- Equivalent releases with different build tags have equal precedence.
- Valid versions sort consistently relative to malformed versions.
- The newest installed version is selected when multiple manifests exist.
- Existing manifest-ID encoding is preserved without migration.
- Regression tests cover supported formats and edge cases.

### PR DESCRIPTION
**Pull request by bobtista**
## Summary

Fix inconsistent publisher-version ordering by routing comparisons through a version scheme declared by each provider.

Generals Online versions previously treated `MMDDYY` values as sortable integers. For example, `060526_QFE1` represents June 5, 2026 and must sort after `121525_QFE1`, which represents December 15, 2025.

## Changes

- Add shared version-scheme abstractions and provider-based scheme resolution.
- Implement numeric, ISO-date, and MMDDYY-QFE schemes.
- Declare the applicable scheme in each provider definition.
- Route update checks, download badges, profile orchestration, and reconciliation through the shared comparer.
- Order Generals Online releases by date and QFE while ignoring build tags for precedence.
- Select the newest installed Generals Online version when multiple manifests exist.
- Keep complete raw versions on manifests.
- Preserve existing Generals Online manifest-ID encoding and extended-version fallback behavior.
- Remove superseded comparison implementations.

## Testing

- Focused version, update-service, download-badge, and reconciliation tests.
- Coverage for cross-year dates, multi-digit QFE values, build tags, malformed versions, culture invariance, and numeric date expansion through 2030.
- GitHub Actions Linux build passed.
- GitHub Actions Windows build passed.
- Build Summary passed.

## Risks and rollback

- Version ordering changes for configured providers are intentional; raw version values remain unchanged.
- Existing Generals Online manifest IDs remain compatible, so no migration is required.
- Unknown configured schemes log a warning and fall back to numeric ordering.
- Rollback consists of reverting this PR; no persisted-data conversion is introduced.

## Related issues

Closes #306

Supersedes #296

### CODE REVIEW

File: GenHub/GenHub.Core/Helpers/GameVersionHelper.cs [hunk 1]
```diff
@@ -122,68 +125,33 @@ public static int NormalizeVersion(string? version)
...
-            return (dateValue * 10) + parsed.Value.Qfe;
+            return ExtractVersionFromVersionString(version);
         }
 
-        // Fallback: use ExtractVersionFromVersionString which handles overflow
-        return ExtractVersionFromVersionString(version);
+        var components = parsed.Components;
+        var mmddyy = (int)((components[1] * 10000) + (components[2] * 100) + (components[0] % 100));
+
+        return (mmddyy * 10) + (int)components[3];
```
[the remark this answers is not in the document]
Comment (undead2146) [reply GenHub/GenHub.Core/Helpers/GameVersionHelper.cs]: ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)

File: GenHub/GenHub.Core/Models/Content/ContentVersion.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ContentVersion"/> struct.
+    /// </summary>
+    /// <param name="components">The version components, most-significant first.</param>
+    public ContentVersion(params long[] components) => _components = components;
+
+    /// <summary>
+    /// Gets the version components, most-significant first.
+    /// </summary>
+    public IReadOnlyList<long> Components => _components ?? NoComponents;
```
[the remark this answers is not in the document]
Comment (undead2146) [reply GenHub/GenHub.Core/Models/Content/ContentVersion.cs]: ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)

File: GenHub/GenHub.Core/Services/Providers/VersionSchemes/MmddyyQfeVersionScheme.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            return false;
+        }
+
+        if (!DateTime.TryParseExact(
+                segments[0],
+                GeneralsOnlineConstants.VersionDateFormat,
+                CultureInfo.InvariantCulture,
+                DateTimeStyles.None,
+                out var date))
```
[the remark this answers is not in the document]
Comment (undead2146) [reply GenHub/GenHub.Core/Services/Providers/VersionSchemes/MmddyyQfeVersionScheme.cs:48]: ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)

File: GenHub/GenHub.Core/Services/Providers/VersionSchemes/NumericVersionScheme.cs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        }
+
+        if (isNumeric1 && hasDot2 && OutranksDateStamp(version2, numeric1))
+        {
+            return -1;
+        }
+
+        if (hasDot1 || hasDot2)
+        {
+            return CompareSegments(version1, version2);
```
[the remark this answers is not in the document]
Comment (undead2146) [reply GenHub/GenHub.Core/Services/Providers/VersionSchemes/NumericVersionScheme.cs]: ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)

File: GenHub/GenHub.Core/Services/Providers/VersionSchemes/NumericVersionScheme.cs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        if (normalized.Length == 10 && normalized[4] == '-' && normalized[7] == '-')
+        {
+            normalized = normalized.Replace("-", string.Empty);
+        }
+
+        return normalized;
+    }
+
+    private static long ExpandShortDate(long version) =>
+        version is >= ShortDateFloor and <= ShortDateCeiling ? CenturyOffset + version : version;
```
[the remark this answers is not in the document]
Comment (undead2146) [reply GenHub/GenHub.Core/Services/Providers/VersionSchemes/NumericVersionScheme.cs]: ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)

File: GenHub/GenHub.Core/Helpers/GameVersionHelper.cs [hunk 6]
```diff
@@ -122,68 +122,57 @@ public static int NormalizeVersion(string? version)
...
+            return (mmddyy * 10) + qfe;
         }
-
-        var parsed = ParseGeneralsOnlineVersion(version);
-        if (parsed != null)
+        catch (ArgumentOutOfRangeException)
         {
-            var dateValue = int.Parse(parsed.Value.Date.ToString("MMddyy"));
-            return (dateValue * 10) + parsed.Value.Qfe;
+            return ExtractVersionFromVersionString(version);
```
[the remark this answers is not in the document]
Comment (undead2146) [reply GenHub/GenHub.Core/Helpers/GameVersionHelper.cs:180]: ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)

**Comment by undead2146** [comment]
🤖 **OpenClaw task queued**

Task ID: `openclaw-1785155503373`

Another task is currently running on this PR. Your task will be processed when the current task completes.

Commands:
- Address code review feedback from @undead2146:

1. GenHub/GenHub.Core/Helpers/GameVersionHelper.cs:154 - _🗄️ Data Integrity & Integration_ | _🟠 Major_ | _⚡ Quick win_  **Preserve multi-digit QFE values in manifest IDs.**  The fixed `* 10` multiplier corrupts valid multi-digit QFEs: `101525_QFE10` becom
2. GenHub/GenHub.Core/Models/Content/ContentVersion.cs:24 - _🎯 Functional Correctness_ | _🟠 Major_ | _⚡ Quick win_  **Defensively copy and hide the backing array.**  Callers can mutate the source array after construction, or cast `Components` back to `long[]
3. GenHub/GenHub.Core/Services/Providers/VersionSchemes/MmddyyQfeVersionScheme.cs:48 - _🎯 Functional Correctness_ | _🟠 Major_ | _⚡ Quick win_  <details> <summary>🧩 Analysis chain</summary>  🏁 Script executed:  ```shell #!/bin/bash set -eu fd -a -e csproj -e props GenHub | xargs -r r
4. GenHub/GenHub.Core/Services/Providers/VersionSchemes/NumericVersionScheme.cs:99 - _🎯 Functional Correctness_ | _🟠 Major_ | _⚡ Quick win_  **Compare normalized dotted versions.**  `TryParse` accepts `release-`/`weekly-` prefixes, but this branch passes the raw strings to `CompareS
5. GenHub/GenHub.Core/Services/Providers/VersionSchemes/NumericVersionScheme.cs:155 - _🎯 Functional Correctness_ | _🟠 Major_ | _⚡ Quick win_  **Preserve six-digit input width before expanding dates.**  `long` parsing removes leading zeroes, so documented YYMMDD values such as `010126
6. GenHub/GenHub.Core/Helpers/GameVersionHelper.cs:198 - _🗄️ Data Integrity & Integration_ | _🟠 Major_ | _⚡ Quick win_  **Reject malformed and overflowing QFE values before deriving manifest IDs.**  `Replace("QFE", ...)` accepts values such as `QFEQFE2`, 
7. GenHub/GenHub.Core/Services/Providers/VersionSchemes/NumericVersionScheme.cs:14 - _🎯 Functional Correctness_ | _🟠 Major_ | _⚡ Quick win_  **Gate the dotted-version override on a validated date.**  `OutranksDateStamp` treats every numeric value above `100000` as a date. Thus `Comp
8. GenHub/GenHub.Core/Helpers/GameVersionHelper.cs:154 - ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)
9. GenHub/GenHub.Core/Models/Content/ContentVersion.cs:24 - ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)
10. GenHub/GenHub.Core/Services/Providers/VersionSchemes/MmddyyQfeVersionScheme.cs:48 - ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)
11. GenHub/GenHub.Core/Services/Providers/VersionSchemes/NumericVersionScheme.cs:99 - ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)
12. GenHub/GenHub.Core/Services/Providers/VersionSchemes/NumericVersionScheme.cs:155 - ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)

File: GenHub/GenHub.Core/Services/Providers/VersionSchemes/NumericVersionScheme.cs [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+namespace GenHub.Core.Services.Providers.VersionSchemes;
+
+/// <summary>
+/// Numeric and semantic versions such as "20251226", "weekly-2025-12-26", "v1.7.2".
+/// Applied to any provider that declares no scheme of its own.
+/// </summary>
+public sealed class NumericVersionScheme : VersionSchemeBase
+{
+    private const long DateStampFloor = 100000;
```
[the remark this answers is not in the document]
Comment (undead2146) [reply GenHub/GenHub.Core/Services/Providers/VersionSchemes/NumericVersionScheme.cs]: ✅ Addressed by OpenClaw (Task: `openclaw-1785154517138`)

File: GenHub/GenHub.Core/Helpers/GameVersionHelper.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bobtista) [reply GenHub/GenHub.Core/Helpers/GameVersionHelper.cs]: Intentionally keeping `101525_QFE10 → 1015260`. On `development`, the previous implementation computes `(101525 * 10) + 10`, so installed manifests may already contain `1015260`. Changing the multiplier to concatenate QFE digits would produce a different ID and require a migration. The regression test pins the actual legacy output; release ordering is handled separately by `MmddyyQfeVersionScheme`.
